src/Microsoft.Diagnostics.Monitoring.Egress/FileSystem/FileSystemEgressProvider.cs

```diff
@@ -0,0 +1,83 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Extensions.Logging;
+using System;
+using System.ComponentModel.DataAnnotations;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Monitoring.Egress.FileSystem
+{
+    internal class FileSystemEgressProvider :
+        EgressProvider<FileSystemEgressProviderOptions, FileSystemEgressStreamOptions>
+    {
+        public FileSystemEgressProvider(FileSystemEgressProviderOptions options, ILogger logger = null)
+            : base(options, logger)
+        {
+        }
+
+        public override async Task<string> EgressAsync(
+            Func<Stream, CancellationToken, Task> action,
+            string name,
+            FileSystemEgressStreamOptions streamOptions,
+            CancellationToken token)
+        {
+            LogAndValidateOptions(name);
+
+            Logger?.LogDebug("Check if directory exists.");
+            if (!Directory.Exists(Options.DirectoryPath))
+            {
+                Logger?.LogDebug("Start creating directory.");
+                Directory.CreateDirectory(Options.DirectoryPath);
+                Logger?.LogDebug("End creating directory.");
+            }
+
+            string targetPath = Path.Combine(Options.DirectoryPath, name);
+
+            if (Options.UseIntermediateFile)
+            {
+                string intermediatePath = Path.Combine(Path.GetTempPath(), Path.GetTempFileName());
```

Comment:
>Path.GetTempPath() [](start = 55, length = 18)

It would be interesting to consider the performance of the move operation here between temp and a mounted volume. For example, can I make the egress faster if I do:
tmp/originalDump -> mountedDrive/intermediate -> mountedDrive/final
as opposed to ->
tmp/originalDump -> tmp/intermediate -> mountedDrive/final

Reply:
(Also this would be very undesirable for mounts that are not emptyDir, such as AzureFiles)

---
In reply to: [518448940](https://github.com/dotnet/diagnostics/pull/1677#discussion_r518448940) [](ancestors = 518448940)

Reply:
Agreed, I'll think of how we can describe that succinctly in the options. I can also do some measurements to see how much impact that may have.

Reply:
Rather than having a UseIntermediateFile options, maybe it is replaced with IntermediateDirectoryPath option, which would imply that the egress provider would write the stream to an arbitrarily named file in the IntermediateDirectoryPath and then move it DirectoryPath. The directory path found in IntermediateDirectoryPath would be created by the egress provider if it does not exist.

Reply:
User can now specify the intermediate directory path with the IntermediateDirectoryPath option on the file system egress options.